### PR TITLE
Adds logic to set eth_call `from` field using `data` field as a fallback.

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

Currently, the WE sets the `from` field on an eth_call request using the current VK. However, as we start supporting multiple VKs, this will not be possible.

We add logic to try and find the `from` field in the eth_call's `data` field, as a fallback.

### What changes were made as part of this PR:

Functional.

- If `eth_call` has missing `from` field, try and determine how to set it based on the `data` field

### CODE REVIEW

File: go/rpcclientlib/viewing_key_client.go [hunk 1]
```diff
@@ -216,14 +221,40 @@ func (c *ViewingKeyClient) addFromAddressToCallParamsIfMissing(method string, ar
...
+	// We iterate over the `data` field, looking for an argument that matches the viewing key address. If we find one,
+	// we set the `from` field to that address.
+	data = data[10:] // We remove the leading "0x" (1 bytes/2 chars) and the method ID (4 bytes/8 chars).
+	for i := 0; i < len(data); i += ethCallPaddedArgLen {
+		if i+ethCallPaddedArgLen > len(data) {
+			return nil, fmt.Errorf("eth_call request did not have its `from` field set, and its `data` field " +
+				"did not contain an address matching a viewing key. Aborting request as it will not be possible to " +
+				"encrypt the response")
+		}
+		// Each encoded parameter in the `data` field is 32 padded bytes. Since an address is 20 bytes, we only want
```
Comment (tudor-malene) [inline go/rpcclientlib/viewing_key_client.go:248]: are you sure this is the general rule?
What if you're sending: "aaaaaaaaa..."- 100 times "a" basically?
Would that be 32 bytes? Is it split up

**Review by tudor-malene — COMMENTED** [review]
looks good in principle.
Have a question

File: go/rpcclientlib/viewing_key_client.go (hunk 1 above)
Comment (joeldudley) [reply go/rpcclientlib/viewing_key_client.go:248]: I believe so, but EthereumDocs is only a second language for me:

```For all types of length at most 32 bytes, the EVENT_INDEXED_ARGS array contains the value directly, padded or sign-extended (for signed integers) to 32 bytes, just as for regular ABI encoding. However, for all “complex” types or types of dynamic length, including all arrays, string, bytes and structs, EVENT_INDEXED_ARGS will contain the Keccak hash of a special in-place encoded value (see [Encoding of Indexed Event Parameters](https://docs.soliditylang.org/en/develop/abi-spec.html#indexed-event-encoding)), rather than the encoded value directly. This allows applications to efficiently query for values of dynamic-length types (by setting the hash of the encoded value as the topic), but leaves applications unable to decode indexed values they have not queried for. For dynamic-length types, application developers face a trade-off between fast search for predetermined values (if the argument is indexed) and legibility of arbitrary values (which requires that the arguments not be indexed). Developers may overcome this tradeoff and achieve both efficient search and arbitrary legibility by defining events with two arguments — one indexed, one not — intended to hold the same value.``` - https://docs.soliditylang.org/en/develop/abi-spec.html

Comment (tudor-malene) [reply go/rpcclientlib/viewing_key_client.go:248]: this is really hard to parse.

Comment (BedrockSquirrel) [reply go/rpcclientlib/viewing_key_client.go:248]: This method might be good to have tests for perhaps? Somewhere we can add edge-cases as we discover them haha

File: go/rpcclientlib/viewing_key_client.go [hunk 2]
```diff
@@ -216,14 +221,40 @@ func (c *ViewingKeyClient) addFromAddressToCallParamsIfMissing(method string, ar
...
+	}
+	data, ok := callParams[reqJSONKeyData].(string)
+	if !ok {
+		return nil, fmt.Errorf("eth_call request's `data` field was not of the expected type `string`")
+	}
 
+	// We iterate over the `data` field, looking for an argument that matches the viewing key address. If we find one,
+	// we set the `from` field to that address.
+	data = data[10:] // We remove the leading "0x" (1 bytes/2 chars) and the method ID (4 bytes/8 chars).
+	for i := 0; i < len(data); i += ethCallPaddedArgLen {
```
Comment (BedrockSquirrel) [inline go/rpcclientlib/viewing_key_client.go:242]: Maybe it would be easier to follow it if was formatted to loop over args like a `while hasNext`, e.g.

```
for arg := getNextArg(data); nextArg != nil; {
   // check arg for address
   ...
    arg = getNextArg(data)
}
```

**Review by BedrockSquirrel — COMMENTED** [review]
Minor refactor suggestion.

And perhaps a further small refactor suggestion: if you had `FindFirstAddr(callData)` then we could test our edgecase scenarios on that quite easily

File: go/rpcclientlib/viewing_key_client.go (hunk 2 above)
Comment (joeldudley) [reply go/rpcclientlib/viewing_key_client.go:243]: I'm splitting it into steps - one to extract all the args, and one to process each arg in turn.

File: go/rpcclientlib/viewing_key_client.go (hunk 1 above)
Comment (joeldudley) [reply go/rpcclientlib/viewing_key_client.go]: It's tested by the wallet extension tests. Do you mean failure tests?

**Review by BedrockSquirrel — APPROVED** [review]
LGTM
